datamap/mv/core/src/main/scala/org/apache/carbondata/mv/datamap/MVHelper.scala

```diff
@@ -596,12 +596,20 @@ object MVHelper {
         case g: GroupBy =>
           MVHelper.updateDataMap(g, rewrite)
       }
-      // TODO Find a better way to set the rewritten flag, it may fail in some conditions.
-      val mapping =
-        rewrittenPlan.collect { case m: ModularPlan => m } zip
-        updatedDataMapTablePlan.collect { case m: ModularPlan => m }
-      mapping.foreach(f => if (f._1.rewritten) f._2.setRewritten())
 
+      // Use coarse equal to set the rewritten flag
+      // We may change the ModularPlan relation
+      // So we match the node excluding child comparison
+      updatedDataMapTablePlan.map(plan => {
+        if (rewrittenPlan.find (
+          oriPlan => oriPlan.rewritten && (plan match {
+            case select: Select => select.coarseEqual(oriPlan)
```

Comment:
It will unnecessarily add the processing to check the equals of all relations here. It is better we can remove this code completely and add the following.

MVHelper
```
def rewriteWithMVTable(rewrittenPlan: ModularPlan, rewrite: QueryRewrite): ModularPlan = {
    if (rewrittenPlan.find(_.rewritten).isDefined) {
      val updatedDataMapTablePlan = rewrittenPlan transform {
        case s: Select =>
          MVHelper.updateDataMap(s, rewrite)
        case g: GroupBy =>
          MVHelper.updateDataMap(g, rewrite)
      }
      updatedDataMapTablePlan
    } else {
      rewrittenPlan
    }
  }
```
Add new method in org.apache.carbondata.mv.plans.modular.GroupBy
```
override def makeCopy(newArgs: Array[AnyRef]): GroupBy = {
    val groupBy = super.makeCopy(newArgs).asInstanceOf[GroupBy]
    if (rewritten) groupBy.setRewritten()
    groupBy
  }
```
 Add new method in org.apache.carbondata.mv.plans.modular.Select
```
  override def makeCopy(newArgs: Array[AnyRef]): Select = {
    val select = super.makeCopy(newArgs).asInstanceOf[Select]
    if (rewritten) select.setRewritten()
    select
  }
```

Reply:
The above can copy the rewritten flag during transformation itself so need to do equals check.

Reply:
done